karoo_gp/test/test_population.py

```diff
@@ -1,145 +1,122 @@
+        assert len(depths) > 1  # There are different depths
+        assert len(n_childrens) > 1  # There are different numbers of children
+    elif tree_type == 'r':
+        count = dict(f=0, g=0)
+        depths = set()
+        for t in population.trees:
+            count[t.tree_type] += 1
+            depths.add(t.depth)
+        _pop, _depth = kwargs['tree_pop_max'], kwargs['tree_depth_base']
+        n_cycles = _pop // (2 * _depth)
+        n_extra = _pop - n_cycles * (2 * _depth)
+        assert count['f'] == n_cycles * _depth  # One 'full' tree at each depth
+        assert count['g'] == count['f'] + n_extra  # That plus extras 'grow'
+        assert len(depths) == kwargs['tree_depth_base']  # Trees at each depth
 
 @pytest.fixture
 def default_evolve_params():
     return dict(
         swim='p',
-        tree_depth_min=1,
+        tree_depth_min=3,
+        tree_depth_max=5,
         evolve_repro=0.1,
         evolve_point=0.1,
         evolve_branch=0.2,
         evolve_cross=0.6,
         tourn_size=7,
     )
 
-def sigfig_round(x, sigfig):
-    """Round to significant figures
-
-        sigfig_round(1111.0, 3) -> 1110
-        sigfig_round(0.1111, 3) -> 0.111
+def test_population_class(tmp_path, paths, default_pop_kwargs,
+                          default_evolve_params):
+    dataset_params = load_data(tmp_path, paths, 'r')
+    terminals = Terminals(dataset_params['terminals'])
+    X, y = dataset_params['X'], dataset_params['y']
 
-    Karoo's expression parser (numpy or tensorflow) uses float32s by default,
-    which are accurate to 7 significant figures. The other digits are artifacts
-    of how python stores floats, i.e. as fractions. The model should only
-    return the significant figures of a calculation. This will be udpated in
-    future version, but this is a workaround so the tests don't fail on GitHub.
-    """
-    return round(x, sigfig -int(np.floor(np.log10(abs(x)))) - 1)
-
-@pytest.mark.parametrize('kernel', ['c', 'r', 'm'])
-def test_population_class(default_kwargs, default_evaluate_params,
-                          default_evolve_params, tmp_path, paths, kernel):
-    # Load the dataset for kernel
-    np.random.seed(1000)
-    tf.set_random_seed(1000)
-    dataset_params = load_data(tmp_path, paths, kernel)
-    terminals = Terminals(dataset_params['terminals'][:-1])
-    functions = Functions([f[0] for f in dataset_params['functions']])
-
-    # Initialize population using dataset terminals
-    kwargs = {
-        **default_kwargs,
-        'tree_pop_max': 10,
-        'terminals': terminals,
-        'functions': functions,
-        'fitness_type': dataset_params['fitness_type']
-    }
-    population = Population.generate(**kwargs)
+    pop_kwargs = dict(**default_pop_kwargs)
+    pop_kwargs['terminals'] = terminals
+    population = Population.generate(**pop_kwargs)
+    assert population.gen_id == 1
+    assert population.fittest().id == 1  # If not evaluated, return first
```

Comment:
In a previous version I raised an error if `fittest()` was called before `fit()`, but

Before calling `fit()` on the model, you should be able to call `predict(X)` and get back a prediction, though you'd expect that prediction to be pretty bad (random). `model.predict(X)` would then call `model.population.fittest()` to determine which tree to use, so you'd want `fittest()` to return *something*, even if `fit()` hasn't been called.

An alternative is to have `model.predict()` check if it's been evaluated, call `fittest()` if so, else pick the first tree. But I think it should return a prediction.

Reply:
Can't `.fittest` still raise and have `.predict` catch the exception and pick the first tree itself, rather than relying on `.fittest` to return it?
Having `.fittest` returning either the fittest tree or a random tree that is not the fittest seems error prone -- what if I forget to call `.fit` and then call `.fittest` and get back a random tree thinking it's the fittest one and keep using that?

Reply:
Yeah, makes sense (done)